File: C++/Standard-Algorithms/MergeSort.cpp
#include<bits/stdc++.h>
using namespace std;
// Time-Complixity: O(N Log N) in all cases
// while merging last one [2,4,5] [1,3] remember - 2 is low,5 is mid,1 is mid+1 and 3 is high
void merge(vector<int> &nums,int lb,int mid,int ub)
{
    int n1=mid-lb+1;
    int n2=ub-mid;

    int L[n1],U[n2];
    for(int i=0;i<n1;i++)
    L[i]=nums[lb+i];

    for(int j=0;j<n2;j++)
    U[j]=nums[mid+1+j];

    int i=0,j=0,k=lb;

    while(i<n1 && j<n2)
    {
        if(L[i]<U[j])
        {
            nums[k]=L[i];
            i++;
        }
        else
        {
            nums[k]=U[j];
            j++;
        }
        k++;
    }

    while(i<n1)
    {
        nums[k]=L[i];
        i++;
        k++;
    }

    while(j<n2)
    {
        nums[k]=U[j];
        j++;
        k++;
    }

}
void mergeSort(vector<int> &nums,int lb,int ub)
{
    if(lb<ub)
    {
        int mid=(lb+ub)/2;
        mergeSort(nums,lb,mid);
        mergeSort(nums,mid+1,ub);
        merge(nums,lb,mid,ub);
    }
}

int main()
{
    vector<int> nums={4,2,5,1,3};
    mergeSort(nums,0,nums.size()-1);

    for(auto no:nums)
    cout<<no<<endl;
}

File: Java/Standard_Algorithms/MergeSort.java
import java.util.*;
class MergeSort
{
    public int nums[]={4,5,1,3,2};

    public void merge(int lower,int mid,int upper)
    {
        int n1=mid-lower+1;
        int n2=upper-mid;

        int L[]=new int[n1];
        int U[]=new int[n2];

        for(int i=0;i<n1;i++)
        L[i]=nums[lower+i];

        for(int j=0;j<n2;j++)
        U[j]=nums[mid+1+j];

        int i=0,j=0,k=lower;

        while(i<n1 && j<n2)
        {
            if(L[i]<U[j])
            {
                nums[k]=L[i];
                i++;
                k++;
            }
            else
            {
                nums[k]=U[j];
                j++;
                k++;

            }
        }

        while(i<n1)
        {
            nums[k]=L[i];
            i++;
            k++;

        }

        while(j<n2)
        {
            nums[k]=U[j];
            j++;
            k++;
        }


    }
    public void mergeSort(int lower,int upper)
    {
        if(lower<upper)
        {
            int mid=(lower+upper)/2;
            mergeSort(lower,mid);
            mergeSort(mid+1,upper);

            merge(lower,mid,upper);
        }
    }

    public void display()
    {
        for(int i:nums)
        System.out.println(i);
    }
    public static void main(String args[])
    {
        MergeSort obj=new MergeSort();
        obj.mergeSort(0,4);
        obj.display();
    }
}